### DISCUSSION TITLE
Pull matching data by sending a list as an argument

### DISCUSSION THREAD

**Discussion by sumchans** [Azure/data-api-builder#1445] [Q&A]
I am not sure if this would be possible using the data-api-builder as of now. I want to retrieve data from the database by sending a list as an argument, an example of a such a list would be `["4161231234","7202228799"]`. 

I have a query like below which works, now I want to send the above array to the query and retrieve data that contains each of the item on the array.
```
query filteritems {
    items(filter: { userPhoneNumber: { contains: "0000" } }) {
        items {
            partitionKey
            userPhoneNumber
            userDisplayName
        }
    }
}
```

An example of such a query on Postgres is -
`SELECT user_id FROM users WHERE user_id LIKE ANY(contacts);`

In the above statement the contacts will have the list of phone numbers.

**Comment by aaronpowell** [comment]
Presently there isn't a way to provide an array such as that for the filter, but you could achieve it slightly differently by using the `or` filter:

```graphql
query filterItems {
  items(filter: { or: [{ userPhoneNumber: { eq: "0000" } } { userPhoneNumber: { eq: "1111" } }]) {
    item {
      partitionKey
      userPhoneNumber
      userDisplayName
    }
  }
}
```

(The filter is written from memory, it should be something close to that if I didn't get it right)

Since you probably have a dynamically sizes list of phone numbers, you're best building the filter in code and passing it as a variable. I'll just JS as an example:

```js
const phoneNumbers = [ /* some phone numbers */ ];

const filters = phoneNumbers.map(phoneNumber => ({ userPhoneNumber: { eq: phoneNumber } }));

const filter = { or: filters };

const query = `
query filterItems ($filter: itemsQueryFilter!) {
  items(filter: $filter) {
    items {
      partitionKey
      ...
    }
  }
}
`;

const results = await execGraphQLQuery(query, { variables: { filter } });
```

This is a little pseudocode esq, but hopefully you can see the intent.

**Reply by sumchans** [reply]
Thanks @aaronpowell. This looks like a solution to me. I still got to try it out.

**Reply by sumchans** [reply]
Hello @aaronpowell,

I just don't understand this line `query filterItems ($filter: itemsQueryFilter!)` What is itemsQueryFilter here?

**Reply by aaronpowell** [reply]
The GraphQL types generated will contain an input type that is generated for the filter of the query.

In this sample I've used it as the type of the variable provided to the query when it's called from the client. See https://graphql.org/learn/queries/#variables on how variables are used, and then it'll be dependent on the client library as to how you provide them.

**Reply by sumchans** [reply]
I converted your javascript code to dart and below is the error I get - 

```
I/flutter (11573): Query failed: Error on line 2, column 21: Expected an argument name
I/flutter (11573):   ╷
I/flutter (11573): 2 │   query filterItems($filter: itemsQueryFilter!) {
I/flutter (11573):   │                     ^
I/flutter (11573):   ╵
```

This is my complete dart code -

```
final phoneNumbers = ["4160000000", "7780000000"];

    final filters = phoneNumbers.map((phoneNumber) => phoneNumber).toList();

    Map<String, dynamic> filter = {
      "or": filters,
    };

    const String getAzureContacts = """ {
      query filterItems (\$filter: itemsQueryFilter!) {
        items(filter: \$filter) {
          items {
            partitionKey
            userPhoneNumber
            userDisplayName
          } 
        }
      }
    }
    """;

    HttpLink link = GlobalVariables().graphqlEndpoint;
    GraphQLClient graphQlClient = GraphQLClient(
      link: link,
      cache: GraphQLCache(
        store: InMemoryStore(),
      ),
    );
    try {
      QueryResult queryResult = await graphQlClient.query(
        QueryOptions(
            fetchPolicy: FetchPolicy.networkOnly,
            document: gql(
              graphql_string.getAzureContacts,
            ),
            variables: {'filter': filter}),
      );
      print(queryResult.data);
    } catch (e) {
      print('Query failed: $e');
    }
```

**Reply by aaronpowell** [reply]
I'm not familiar with Dart, but a quick web search shows me https://pub.dev/packages/graphql#query as a sample query with a variable

**Reply by sumchans** [reply]
Thanks @aaronpowell
